### PR TITLE
Replace the old scraping code with new WebFlux-based scraping actions

### PR DESCRIPTION
**Pull request by mcpierce**
## Status
**READY**

## Migrations
YES

## Description
Designed a new scraping system based on a ScrapingAdaptor and individual ScrapingAction objects. Replaced the old caching tables with a new concise pair of tables.

### CODE REVIEW

File: comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+   * Returns a list of volumes for the given series name.
+   *
+   * @param apiKey the API key
+   * @param seriesName the series name
+   * @param skipCache the skip cache flag
+   * @return the list of volumes
+   * @throws ScrapingException if an error occurs
+   */
+  List<ScrapingVolume> getVolumes(String apiKey, String seriesName, boolean skipCache)
```
Comment (BRUCELLA2) [inline comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java:43]: Parameters can be declared "final" (they are final in the implementation).

File: comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+   * Returns a single issue.
+   *
+   * @param apiKey the API key
+   * @param volume the volume
+   * @param issueNumber the issue number within the volume
+   * @param skipCache the skip cache flag
+   * @return the issue or null
+   * @throws ScrapingException if an error occurs
+   */
+  ScrapingIssue getIssue(String apiKey, Integer volume, String issueNumber, boolean skipCache)
```
Comment (BRUCELLA2) [inline comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java:56]: Parameters can be declared "final" (they are final in the implementation).

File: comixed-scrapers/src/main/java/org/comixed/scrapers/comicvine/actions/ComicVineGetIssueDetailsAction.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    this.addField("location_credits");
+    this.addField("story_arc_credits");
+    this.addField("person_credits");
+
+    if (StringUtils.isEmpty(this.getApiKey())) throw new ScrapingException("Missing API key");
+    if (StringUtils.isEmpty(this.issueId)) throw new ScrapingException("Missing issue id");
+
+    log.debug(
+        "Querying ComicVine for issue: id={} API key={}", this.issueId, this.getMaskedApiKey());
+    final String url = this.createUrl(this.baseUrl, this.getEndpoint());
```
Comment (BRUCELLA2) [inline comixed-scrapers/src/main/java/org/comixed/scrapers/comicvine/actions/ComicVineGetIssueDetailsAction.java:68]: if baseUrl is null couldn't that be a problem?

**Review by BRUCELLA2 — CHANGES_REQUESTED** [review]
3 minor details to check.

File: comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java (hunk 1 above)
Comment (mcpierce) [reply comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java:43]: Isn’t the final keyword only useful on an implementation: ie an implementing class can drop the final keyword?

File: comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java (hunk 2 above)
Comment (mcpierce) [reply comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java:56]: See previous reply.

File: comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java (hunk 1 above)
Comment (BRUCELLA2) [reply comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java:43]: I learn something :) I'm still a rookie ;)
It's true, final can be drop in the implementation and adding the final word in an interface can therefore be counter-intuitive as the interface contract that indicates that the parameter will not be reassigned could be reassigned in the implementation.
Don't you think then that we should remove the "final" in the scrapeComic method in the interface?

File: comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java (hunk 2 above)
Comment (BRUCELLA2) [reply comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java:56]: ok see previous reply.

File: comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java (hunk 1 above)
Comment (mcpierce) [reply comixed-scrapers/src/main/java/org/comixed/scrapers/adaptors/ScrapingAdaptor.java:43]: That's what I was getting at before: "final" only applies to an actual implementation of the method and not to the interface. We could put "final" here and someone could implement the interface and leave it off and it wouldn't be a compile time error: final serves no purpose in an interface.

File: comixed-scrapers/src/main/java/org/comixed/scrapers/comicvine/actions/ComicVineGetIssueDetailsAction.java (hunk 3 above)
Comment (mcpierce) [reply comixed-scrapers/src/main/java/org/comixed/scrapers/comicvine/actions/ComicVineGetIssueDetailsAction.java:68]: Yep, thanks for catching that. I've updated the branch to include at test for that condition.

**Review by BRUCELLA2 — APPROVED** [review]
it's ok for me

**Pull request merged**